File: 0.0.1/layer/dense.py
import numpy as np
from activation_function.relu import Vectorized_ReLu
from activation_function.sigmoid import Vectorized_Sigmoid

class Dense:

    def __init__(self, input_size = 0, layer_size = 0, randomization = True, activation_function = 'ReLu'):
        self.input_size = input_size
        self.layer_size = layer_size
        self.activation_function = activation_function
        if randomization:
            self.weights = 0.1*np.random.randn(input_size, layer_size)
            self.bias = 0.1*np.random.randn(layer_size)
        else:
            self.weights = np.zeros((input_size, layer_size))
            self.bias = np.zeros(layer_size)

    def run(self, input):
        self.value = np.matmul(input, self.weights) + self.bias
        if self.activation_function == 'ReLu':
            self.output = Vectorized_ReLu(self.value)
        elif self.activation_function == 'Sigmoid':
            self.output = Vectorized_Sigmoid(self.value)
        self.propagation_units = np.zeros((self.input_size, self.layer_size, np.shape(input)[0]))

File: 0.2.1/layer/dense.jl
module dense
    mutable struct Dense
        activator::Any

        input_size::Int64
        layer_size::Int64
        activation_function::Module

        weights::Array{Float32}
        biases::Array{Float32}

        value::Array{Float32}
        output::Array{Float32}
        propagation_units::Array{Float32}

        Vdw::Array{Float32}
        Sdw::Array{Float32}
        Vdb::Array{Float32}
        Sdb::Array{Float32}

        function Dense(;input_size::Int64, layer_size::Int64, activation_function::Module, randomization::Bool=true, reload::Bool=false)
            if reload
                return new(activate_Dense)
            end

            if randomization
                weights = 0.1*rand(Float32, (layer_size, input_size))
                biases = 0.1*rand(Float32, layer_size)
            else
                weights = zeros(Float32, (layer_size, input_size))
                biases = zeros(Float32, layer_size)
            end
            new(activate_Dense, input_size, layer_size, activation_function, weights, biases)
        end
    end

    function activate_Dense(layer::Dense, input::Array{Float32})
        layer.value = layer.weights*input .+ layer.biases
        layer.output = layer.activation_function.func(layer.value)
        layer.propagation_units = zeros(Float32, (layer.layer_size, layer.input_size, size(input, 2)))

        layer.Vdw = zeros(Float32, size(layer.weights))
        layer.Sdw = zeros(Float32, size(layer.weights))
        layer.Vdb = zeros(Float32, size(layer.biases))
        layer.Sdb = zeros(Float32, size(layer.biases))
    end
end
